### ISSUE TITLE
Python Atom <--> Handle confusion

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#201]
Some python functions take Handles, some take Atoms, its hard to tell which.   Its is easy to get a Handle from an atom, but hard/confusing to get an Atom from a Handle:
```
h = Handle
a = Atom( h, atomspace)
```

There is no reason to have both an Atom class, and a handle class in the python bindings. The two concepts are identical; one should be removed.

**Comment by AmeBel** [comment]
So which should be the standard?

**Comment by linas** [comment]
Atom, I think.   I see no purpose in having Handle.  It doesn't do anything useful.

**Comment by inflector** [comment]
Using only Atoms will make it easier to understand so I concur with the simplification goal. Two caveats:

1) This is a fair bit of work if I recall, and may not be possible with Cython due to restrictions on the interface between Python and Cython types during Python object creation. There is a reason it was done this way that I understood at one time but which now eludes me. I'm not sure the current implementation was the best way, or the only way, just that it was more rational than it would appear. I believe it had to do with the way that Atoms are created from UUID lists internally.

Best way to find out is to do it.

2) This proposed change will likely break a fair bit of existing Python code, so it's the kind of change that ought to be done with a bit of warning so someone doesn't just pull the latest and have their world break without notice. Perhaps we ought to put a notice on the Google group for Python users and poll existing projects to see how much this will affect their code to give a heads up.

**Comment by AndrewBuck** [comment]
Why isn't there a get_atom_by_handle() function in the atomspace class?  Honestly, any of the functions which take an atom as an argument (or a handle) should just have an overloaded method which takes the alternate version and does the conversion for you, but having a get_atom_by_handle() method makes it possible to do the conversion yourself manually if you want to.

This is something I have had trouble with in working on the minecraft bot where sometimes you do a function call and get the atoms themselves back and sometimes you get handles.  But in theory one should be as good as the other to any of the atomspace routines.

**Comment by AndrewBuck** [comment]
Oh, I should also add that there is (at least in theory) a reason for allowing both atoms and atom handles.  If you are storing or processing large amounts of atom data outside of the atomspace (for example if the bot reads out a bunch of block atoms to look at them all or do some operation to all of them, etc) then you may end up making copies or references to a large number of atoms.  Being able to store or work with atom handles and then only access the atoms themselves for certain atoms would allow memory and possibly computation savings, depending on what you are doing with all those atoms.

**Comment by linas** [comment]
OK, here's the deal: Atoms and Handles are exactly the same thing. Or rather, they  should be. There should be no distinction made between the two.  It was a design error in the cython bindings to treat these two things as if they are different: conceptually, they are not.  The historical reason for the error had to do with a misunderstanding about databases and distribute programming. Oh well.

The impact on the C++ code is this: In C++, class Handle is a pointer to Atom, and it also stores a UUID in it. I want to remove UUID from class Handle (so that class Handle becomes identical to AtomPtr).  The only code that uses that UUID is the cython code; no one else uses it.  (Other code does use UUID's but they use it in a "responsible" fashion, viz only during network operations.)   

Removing the UUID from Handle will simplify a big blob of spaghetti code in the atomspace, and should improve performance, so I can't wait for this change.  It should also open new doors for the protoatom work, so that we can attach attributes to atoms (e.g. simple floats, vectors of floats, strings, etc.) which has been a long-requested feature.

**Comment by linas** [comment]
@AndrewBuck : regarding your comments: You can always ask an atom for it's UUID, and you can always look up an atom according to it's UUID.  I don't recommend it: UUID's are meant for databases and for distributed computing, and most people should not be doing database and distributed programming.  But if you really really need to, that's OK, whatever. I don't really care.

The problem for the C++ code is that the C++ class Handle stores the UUID, and it should not.  As a result, there is a lot of complex code that has to perform various checks and complicated re-writes, to keep everything in sync. It would be nice to remove this code.  Right now, cython is the only code in the whole system that uses it: its the blocker to progress on other fronts.

**Comment by linas** [comment]
btw protoatom is being tracked in #513

**Comment by AndrewBuck** [comment]
@linas ok, I misunderstood the exact nature of the atoms and handles then.  Feel free to ignore my comments, I withdraw my objections as they were made on incorrect assumptions.

**Comment by inflector** [comment]
Linas wrote:
> In C++, class Handle is a pointer to Atom, and it also stores a UUID in it.

I don't see this in the current code. There is a `value` member function:
```
UUID value(void) const;
```
but no member variables of type UUID in the Handle class. And `Handle::value()`  defers to it's Atom:
```
UUID Handle::value(void) const {
    const Atom* a = operator->();
    if (a) return a->getUUID();
    return ULONG_MAX;
}
```
This `Handle::value` code calls operator -> which does a get() on the underlying shared_ptr to get the reference to the atom and then gets the UUID from that atom.

So, perhaps you've already done a partial removal of the UUID at some prior time?

As far at the work for removing Python Handles, there are a lot of API calls which take Handles at the moment which the Cython interface wraps. 

For instance in `opencog/atomspace/opencog/query/BindLinkAPI.h` we have:
```
Handle bindlink(AtomSpace*, const Handle&);
Handle single_bindlink (AtomSpace*, const Handle&);
Handle af_bindlink(AtomSpace*, const Handle&);
TruthValuePtr satisfaction_link(AtomSpace*, const Handle&);
Handle satisfying_set(AtomSpace*, const Handle&);
Handle recognize(AtomSpace*, const Handle&);
```
If we are to remove the Python Handle object from the Python API, then we'll either need to change these or change the API Cython wrappers to use Atom instead of Handle so it no longer directly matches the C++ API. This latter seems like what you're proposing. e.g. for `bindlink` we currently have:
```
def bindlink(AtomSpace atomspace, Handle handle):
    cdef cHandle c_result = c_bindlink(atomspace.atomspace,
                                       deref(handle.h))
    cdef Handle result = Handle(c_result.value())
    return result
```
and the change proposed is to attempt to end up with something this:
```
def bindlink(AtomSpace atomspace, Atom atom):
    cdef cHandle c_result = c_bindlink(atomspace.atomspace, atom.get_handle_reference())
    cdef Atom result = atom_from_c_handle(c_result, atomspace)
    return result
```
so the input and output to the various functions is the Python Atom rather than the current Python Handle.

The removal of the Python Handle object does not preclude Python Atom from using C++ Handle objects internally, in fact, they will need to do so. Currently, a Python Handle object keeps a heap-allocated reference to the C++ Handle, which in turn has a shared_ptr reference to the C++ Atom object itself. If we are to remove the Python Handle object altogether, then this C++ Handle reference will be stored directly in the Python Atom object.

There is a lot of code that wraps HandleSeq iteration as well. So where we currently have:
```
        cdef vector[cHandle].iterator c_handle_iter
        cdef cHandle current_c_handle
        c_handle_iter = handle_vector.begin()

        while c_handle_iter != handle_vector.end():
            current_c_handle = deref(c_handle_iter)
            temp_handle = Handle(current_c_handle.value())
            yield Atom(temp_handle,self)
            inc(c_handle_iter)
```
this will be replaced with:
```
        cdef vector[cHandle].iterator c_handle_iter
        cdef cHandle current_c_handle
        c_handle_iter = handle_vector.begin()

        while c_handle_iter != handle_vector.end():
            yield atom_from_c_handle(deref(c_handle_iter), self)
            inc(c_handle_iter)
```
@linas Does the above match your expectations for the changes proposed here?

**Comment by linas** [comment]
Well, Huh.  I guess I must heave realized that cython was not blocking the UUID work, so I was able to remove it after all.  Interesting; I'd forgotten that.
 
@inflector Well, as you now see -- the C++ Handle and the C++ AtomPtr are almost the same thing.  The C++ Handle class adds assorted convenience methods, but that's it; it doesn't "do anything" per se.

So, for cython, maybe it wold be easier to get rid of Atom, and make everything be Handle?  The point is, you don't need both.  The python Atom class is kind-of hard to work with, anyway, so perhaps we should get rid of that.

What caused me to open this request was that I was writing some python code, and some API wanted me to use atoms instead of handles, and doing the conversions just made everything complicated.  The goal is to make it easy for python coders to write code.  Consistently using Handle everywhere is probably the easiest way to achieve this.

**Comment by linas** [comment]
So, if I `grep "Atom " *` in the pyx files, I see that Atom is used only in a five public API functions:
```
get_atoms_by_target_atom
get_predicates
get_predicates_for
execute_atom
evaluate_atom
```

Change these five places to take Handle instead of Atom, and I think we're done.  After that, I guess you can just remove `atom.pyx`, right? Something like that.

**Comment by inflector** [comment]
Your grep counts the places where `Atom` is a parameter in but misses where `Atom` or a list of `Atom`s is returned as a function result. The helper function `convert_handle_seq_to_python_list` returns a list of Atoms which it constructs from internal C++ Handle vectors. There are 10 places where a call is made to `convert_handle_seq_to_python_list`:
```
cdef convert_handle_seq_to_python_list(vector[cHandle] handles, AtomSpace atomspace):
    cdef vector[cHandle].iterator iter
    cdef cHandle i
    result = []
    iter = handles.begin()
    while iter != handles.end():
        i = deref(iter)
        temphandle = Handle(i.value())
        result.append(Atom(temphandle, atomspace))
        inc(iter)
    return result
```
Further, for each of these 10 calls, there is a corresponding x call, e.g. an `xget_outgoing` using a yield of an `Atom` corresponding to the `get_outgoing` which returns a Python list of `Atom`s. 

Additionally, the bracket operator code for `atomspace` that allows one to lookup atoms by `Handle` using bracket syntax ala:
```
atom = atomspace[targetHandle]
```
returns an `Atom`.

So this change will break a lot of existing Python code. I happen to agree that we should do it, so I'm not writing the above to suggest we do otherwise. I agree that we don't need two separate classes for what is essentially one concept, the Atom.

I just reiterate the case for giving people a heads up on the email list / Google group.

Finally, I suggest we go further and remove the name Handle rather than Atom. As an implementation, I'm suggesting we keep the internals for Handle and rename it Atom, and then make everything work with the new combined Atom semantics. Using the Atom name will make better sense to new users and not require these Python users to have to learn the extra complication of what a Handle is.

Handle is an internal C++ implementation detail rather than a conceptual-level entity. Further, merging Atom and Handle while keeping the name Atom will probably require less change to existing Python code.

**Comment by linas** [comment]
sounds good to me!

**Comment by inflector** [comment]
With Pull Request https://github.com/opencog/atomspace/pull/604 the Python Handle object is no more. 

I'm going to write a note for the list to give Python users a heads up in case they have any issues making the conversion once the request is approved and changes committed to master.

**Comment by inflector** [comment]
@linas Now that Pull Request https://github.com/opencog/opencog/pull/2007 has been merged. This issue should be resolved and ready to be closed.

**Comment by linas** [comment]
OK thanks!
